packages/test/test-utils/src/timeoutUtils.ts

```diff
@@ -61,3 +163,33 @@ export async function timeoutPromise<T = void>(
             });
     });
 }
+
+// Create a promise based on test timeout and the timeout options
+export async function timeoutPromise<T = void>(
+    executor: (resolve: (value: T | PromiseLike<T>) => void, reject: (reason?: any) => void) => void,
+    timeoutOptions: TimeoutWithError | TimeoutWithValue<T> = {},
+): Promise<T> {
+    // create the timeout error outside the async task, so its callstack includes
+    // the original call site, this makes it easier to debug
+    const err = timeoutOptions.reject === false
+        ? undefined
+        : new Error(timeoutOptions.errorMsg ?? "Test timed out");
```

Comment:
should we pass this error into getTimeoutPromise rather than create a different one there?

Reply:
seems strange we'll get different failures with different stack depending on timeout path

Reply:
don't we want to distinguish explicit timeout version "test" timeout?

Reply:
maybe. i assume with this we'll move away from duration based timeout in general